Add Security Profiles Operator Manual Migration Guide

Vincent056 opened the pull request:
#### What type of PR is this?
<!--
/kind documentation
-->

#### What this PR does / why we need it:

This commit introduces a comprehensive guide for manually migrating the Security Profiles Operator (SPO) from namespace-scoped to cluster-scoped Custom Resource Definitions (CRDs).


#### Which issue(s) this PR fixes:

<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.

Fixes #

or

None
-->

#### Does this PR have test?

<!--
If tests aren't applicable just write N/A.
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note

```

Vincent056 committed:
Add Security Profiles Operator Manual Migration Guide

This commit introduces a comprehensive guide for manually migrating the Security Profiles Operator (SPO) from namespace-scoped to cluster-scoped Custom Resource Definitions (CRDs).

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,564 @@
+SELINUX_TYPES=$(oc get selinuxprofiles,rawselinuxprofiles -A -o json | jq -r '.items[].status.usage' | sort -u)
+
+# Find pods using any of these SELinux types
+> $BACKUP_DIR/workload-analysis/selinux-usage.txt
```
This line need to be removed.

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,564 @@
+### Backup Profile Bindings
+```bash
+# Backup all profile bindings
+for binding in $(oc get profilebindings -n openshift-security-profiles -o name); do
````
When backing up all profile bindings, we can not use `-n openshift-security-profiles` here. It is better to use `--all-namespaces` parameter.

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,564 @@
+# Backup all profile bindings
+for binding in $(oc get profilebindings -n openshift-security-profiles -o name); do
+  name=$(echo $binding | cut -d'/' -f2)
+  oc get profilebinding $name -n openshift-security-profiles -o json | jq '.' > \
```
same as above

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,564 @@
+### Backup Security Profiles
+```bash
+# Backup seccomp profiles
+for profile in $(oc get seccompprofiles -n openshift-security-profiles -o name); do
````
why only backup the seccompprofiles in ns openshift-security-profiles here? It is better to use --all-namespaces parameter.

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,564 @@
+# Backup seccomp profiles
+for profile in $(oc get seccompprofiles -n openshift-security-profiles -o name); do
+  name=$(echo $profile | cut -d'/' -f2)
+  oc get seccompprofile $name -n openshift-security-profiles -o json | jq '.' > \
```
same as above

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,564 @@
+done
+
+# Backup SELinux profiles
+for profile in $(oc get selinuxprofiles -n openshift-security-profiles -o name); do
```
same as above

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,564 @@
+# Backup SELinux profiles
+for profile in $(oc get selinuxprofiles -n openshift-security-profiles -o name); do
+  name=$(echo $profile | cut -d'/' -f2)
+  oc get selinuxprofile $name -n openshift-security-profiles -o json | jq '.' > \
```
same as above

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,564 @@
+done
+
+# Backup raw SELinux profiles
+for profile in $(oc get rawselinuxprofiles -n openshift-security-profiles -o name); do
```
same as above

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,564 @@
+# Backup raw SELinux profiles
+for profile in $(oc get rawselinuxprofiles -n openshift-security-profiles -o name); do
+  name=$(echo $profile | cut -d'/' -f2)
+  oc get rawselinuxprofile $name -n openshift-security-profiles -o json | jq '.' > \
```
same as above

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,564 @@
+### Delete Profile Bindings First
+```bash
+# List and delete all profile bindings
+oc get profilebindings -n openshift-security-profiles
````
same as above. Profilebindings can be created in all namespaces
```
% oc get profilebindings -A
NAMESPACE     NAME                      AGE
selinuxtest   profile-binding-selinux   109m
test-spo      busybox-binding           22m
test-spo2     busybox-binding           62m
test-spo2     profile-binding-selinux   41m
```

xiaojiey replied:
One more issue: before delete any profilebinding, the activeWorkloads should be deleted first. Otherwise, the profilebinding deleting will fail.

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,564 @@
+```bash
+# List and delete all profile bindings
+oc get profilebindings -n openshift-security-profiles
+oc delete profilebindings --all -n openshift-security-profiles
````
same as above

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,564 @@
+
+**Actual command and output:**
+```bash
+oc delete profilebinding profile-binding-seccomp profile-binding-selinuxd -n openshift-security-profiles
````
same as above

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,564 @@
+### Delete Security Profiles
+```bash
+# Delete all security profiles
+oc delete seccompprofiles,selinuxprofiles,rawselinuxprofiles --all -n openshift-security-profiles
````
same as above, the seccompprofiles, selinuxprofiles and rawselinuxprofiles can be created in all namespaces.

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,564 @@
+
+
+### Delete Namespace
+```bash
````
Before delete the namespace, you also need to run below command:
`oc delete mutatingwebhookconfigurations spo-mutating-webhook-configuration`

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,564 @@
+
+To uninstall the Security Profiles Operator (SPO) Operator, follow these steps:
+
+### Uninstall via OperatorHub (Recommended)
```
Per the experience, this step `Uninstall via OperatorHub` is not necessary.

Vincent056 committed:
Enhance SPO Manual Migration Guide with namespace-wide backup and deletion commands

This update modifies the backup and deletion sections of the Security Profiles Operator (SPO) migration guide to support operations across all namespaces. It includes commands to back up profile bindings, seccomp profiles, SELinux profiles, and raw SELinux profiles from all namespaces, as well as updated deletion instructions for these resources.

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,617 @@
+Profile bindings backed up:
+profilebindings-openshift-security-profiles-profile-binding-seccomp.json
+profilebindings-openshift-security-profiles-profile-binding-selinuxd.json
+```
````
Do we need to backup profilerecording?

Vincent056 replied:
that's a good point, but I wonder if profile recording is something customer would run at all time, would it be okay if we tell customer they need to check if they have any profile recording in progress, they need to manage those accordingly if so and point them to:
```
**IMPORTANT**: If you have any profile recording in progress, you should also remove them, as well as associated active workloads with the profile recordings.
[Manage Profile Recording](https://docs.redhat.com/en/documentation/openshift_container_platform/4.12/html/security_and_compliance/security-profiles-operator#spo-recording-profiles_spo-seccomp)
```

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,617 @@
+Next, update all workloads to remove references to SPO profiles. You must either recreate pods without these profiles, or update deployments, statefulset, and other workloads to remove the profile references from their pod templates.
+
+
+### Delete Security Profiles
```
Before deleting the security profiles, we need to delete the active workload for the security profiles first.

Vincent056 replied:
Added:
```
Next, update all workloads to remove references to SPO profiles. You must either recreate pods without these profiles, or update deployments, statefulset, and other workloads to remove the profile references from their pod templates.
```
I think it is hard to make recommendations here since it is up to customer to decide how to manage their active workload.

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,617 @@
+oc get seccompprofiles,selinuxprofiles,rawselinuxprofiles -A
+```
+
+## Step 5: Uninstall SPO Operator
````
Before the uninstall, deleting the active workloads for the profilerecording and deleting the profilerecording is also a mandatory step(better before the deleting security profiles step.)

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
````diff
@@ -0,0 +1,617 @@
+
+```bash
+# Delete all SPO CRDs
+for crd in seccompprofiles selinuxprofiles rawselinuxprofiles profilebindings profilerecordings securityprofilenodestatuses securityprofilesoperatordaemons apparmorprofiles
````
Need to be updated to:
`for crd in seccompprofiles selinuxprofiles rawselinuxprofiles profilebindings profilerecordings securityprofilenodestatuses securityprofilesoperatordaemons apparmorprofiles; do`

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,617 @@
+    echo "Restoring: $(basename "$file")"
+    oc apply -f "$file"
+  fi
+done
```
@Vincent056 I failed to restore the security profiles. Could you please take a look? Thanks.
```
echo "Restored profiles:"
oc get seccompprofiles,selinuxprofiles,rawselinuxprofiles
Restoring: rawselinuxprofiles-errorlogger.json
Error from server (NotFound): error when creating "spo-manual-backup-20250826-070749/profiles-to-restore/rawselinuxprofiles-errorlogger.json": the server could not find the requested resource (post rawselinuxprofiles.security-profiles-operator.x-k8s.io)
Restoring: seccompprofiles-log-enricher-trace.json
Error from server (NotFound): error when creating "spo-manual-backup-20250826-070749/profiles-to-restore/seccompprofiles-log-enricher-trace.json": the server could not find the requested resource (post seccompprofiles.security-profiles-operator.x-k8s.io)
Restoring: seccompprofiles-sleep-sh-pod.json
Error from server (NotFound): error when creating "spo-manual-backup-20250826-070749/profiles-to-restore/seccompprofiles-sleep-sh-pod.json": the server could not find the requested resource (post seccompprofiles.security-profiles-operator.x-k8s.io)
Restoring: selinuxprofiles-errorlogger.json
Error from server (NotFound): error when creating "spo-manual-backup-20250826-070749/profiles-to-restore/selinuxprofiles-errorlogger.json": the server could not find the requested resource (post selinuxprofiles.security-profiles-operator.x-k8s.io)
```

xiaojiey replied:
@Vincent056 I am a little confused. I only removed label for rawselinuxprofile. But all profiles restore/creating succeeded on a fresh installed cluster. I think the operator was not installed correctly last time. Maybe something wrong with the previous uninstallation? Let me try again and give you feedback later.
```
$ for file in $BACKUP_DIR/profiles-to-restore/*.json; do
  if [ -f "$file" ]; then
    echo "Restoring: $(basename "$file")"
    oc apply -f "$file"
  fi
done
Restoring: rawselinuxprofiles-errorlogger.json
rawselinuxprofile.security-profiles-operator.x-k8s.io/errorlogger unchanged
Restoring: seccompprofiles-log-enricher-trace.json
seccompprofile.security-profiles-operator.x-k8s.io/log-enricher-trace created
Restoring: seccompprofiles-sleep-sh-pod.json
seccompprofile.security-profiles-operator.x-k8s.io/sleep-sh-pod unchanged
Restoring: selinuxprofiles-errorlogger.json
selinuxprofile.security-profiles-operator.x-k8s.io/errorlogger created
Restoring: selinuxprofiles-nginx-secure.json
selinuxprofile.security-profiles-operator.x-k8s.io/nginx-secure created
Restoring: selinuxprofiles-spo-recording-hello-openshift-5n58t.json
selinuxprofile.security-profiles-operator.x-k8s.io/spo-recording-hello-openshift-5n58t created
Restoring: selinuxprofiles-spo-recording-hello-openshift-c9tf8.json
selinuxprofile.security-profiles-operator.x-k8s.io/spo-recording-hello-openshift-c9tf8 created
Restoring: selinuxprofiles-spo-recording-hello-openshift-ms6qg.json
selinuxprofile.security-profiles-operator.x-k8s.io/spo-recording-hello-openshift-ms6qg created
Restoring: selinuxprofiles-spo-recording-hello-openshift2-5n58t.json
selinuxprofile.security-profiles-operator.x-k8s.io/spo-recording-hello-openshift2-5n58t created
Restoring: selinuxprofiles-spo-recording-hello-openshift2-c9tf8.json
selinuxprofile.security-profiles-operator.x-k8s.io/spo-recording-hello-openshift2-c9tf8 created
Restoring: selinuxprofiles-spo-recording-hello-openshift2-ms6qg.json
selinuxprofile.security-profiles-operator.x-k8s.io/spo-recording-hello-openshift2-ms6qg created
```

Vincent056 committed:
Update SPO Manual Migration Guide with important notes on profile recordings and metadata cleanup

This commit adds a note regarding the removal of active workloads and ongoing profile recordings during the migration process. Additionally, it enhances the cleanup script by including the deletion of metadata fields such as annotations, labels, managed fields, and owner references from backup profiles.

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,625 @@
+    echo "  Original namespace: $namespace"
+    
+    # Prepare the profile (remove namespace and clean metadata)
+    jq 'del(.metadata.namespace) | 
```
Need to add the "|" pipe operator at the end of each jq command:
```
    jq 'del(.metadata.namespace) | 
        del(.metadata.resourceVersion) | 
        del(.metadata.uid) | 
        del(.metadata.creationTimestamp) | 
        del(.metadata.generation) | 
        del(.metadata.annotations) |
        del(.metadata.labels) |
        del(.metadata.managedFields) |
        del(.metadata.ownerReferences) |
        del(.status) | 
        del(.metadata.finalizers)' "$file" > "$BACKUP_DIR/profiles-to-restore/${profile_type}-${name}.json"
```

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,625 @@
+    image: nginx:1.19.1
+    securityContext:
+      seLinuxOptions:
+        type: errorlogger-selinuxd-test.process
```
It should be `type: errorlogger-selinuxd-test_.process`

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,625 @@
+  - name: nginx
+    securityContext:
+      seLinuxOptions:
+        type: errorlogger-selinuxd-test.process  # No namespace in type
```
It should be `type: errorlogger-selinuxd-test_.process` here

xiaojiey commented on doc/SPO-MANUAL-MIGRATION-GUIDE.md:
```diff
@@ -0,0 +1,625 @@
+seccompprofile.security-profiles-operator.x-k8s.io/profile-block-all   Installed   30s
+
+NAME                                                                           USAGE                                            STATE
+selinuxprofile.security-profiles-operator.x-k8s.io/errorlogger-selinuxd-test  errorlogger-selinuxd-test.process                Installed
```
It should be `errorlogger-selinuxd-test_.process`
